torch/_inductor/fx_passes/micro_pipeline_tp.py

```diff
@@ -0,0 +1,441 @@
+import operator
+from dataclasses import dataclass
+from typing import cast, List, Set, Tuple, Union
+
+import torch
+
+from ..pattern_matcher import (
+    CallFunction,
+    Ignored,
+    KeywordArg,
+    ListOf,
+    MULTIPLE,
+    PatternMatcherPass,
+    register_graph_pattern,
+)
+
+aten = torch.ops.aten
+c10d = torch.ops._c10d_functional
+patterns = PatternMatcherPass()
+
+
+def _is_backward(graph: torch.fx.Graph) -> bool:
+    placeholders = []
+    for node in graph.nodes:
+        if node.op != "placeholder":
+            break
+        placeholders.append(node)
+    return not all(node.name.startswith("primal") for node in placeholders)
+
+
+def _compute_mm_arithmetic_intensity(M: int, N: int, K: int) -> float:
+    return M * N * K / (M * K + N * K + M * N)
+
+
+def _filter_nodes_by_target(nodes: List[torch.fx.Node], target) -> List[torch.fx.Node]:
+    return [x for x in nodes if x.target == target]
+
+
+def _find_ancestors(node: torch.fx.Node) -> Set[torch.fx.Node]:
+    ancestors = set()
+    ancestors.add(node)
+    cur_nodes = [node]
+    while len(cur_nodes) > 0:
+        new_nodes = []
+        for node in cur_nodes:
+            for inp in node.all_input_nodes:
+                if inp not in ancestors:
+                    ancestors.add(inp)
+                    new_nodes.append(inp)
+        cur_nodes = new_nodes
+    return ancestors
+
+
+def _can_schedule_y_before_x(
+    x: torch.fx.Node, y: torch.fx.Node
+) -> Tuple[bool, Set[torch.fx.Node]]:
+    """
+    Check if y can be reordered before x and return the ancestors of y
+    (inclusive).
+    """
+    y_ancestors = _find_ancestors(y)
+    if x in y_ancestors:
+        return False, y_ancestors
+
+    return True, y_ancestors
+
+
+@dataclass
+class _2DMatmul:
+    node: torch.fx.Node
+    B_node: torch.fx.Node
+    B_node_ancestors: Set[torch.fx.Node]
+
+    def replace_with(self, new_node: torch.fx.Node) -> None:
+        """
+        Replace the matmul with the new node.
+        """
+        self.node.replace_all_uses_with(new_node)
+
+
+@dataclass
+class _NDMatmul:
+    nodes: List[torch.fx.Node]
+    B_node: torch.fx.Node
+    B_node_ancestors: Set[torch.fx.Node]
+
+    def replace_with(self, new_node: torch.fx.Node) -> None:
+        """
+        Replace the matmul with the new node.
+
+        ND-matmul is a sequence of reshape -> mm -> reshape in the graph. The
+        second reshape node is replaced with `new_node`.
+
+        In addition, we ensure that the original mm node ends up with zero
+        users by replacing it with a reverse reshape of `new_node`.
```

Comment:
could you elaborate this part more? how could replacing it with a reverse shape of `new_node` results in original mm node with zero users?

Reply:
An ND-matmul shows up in fx graphs as `reshape -> mm -> reshape` sequences. The first reshape flattens the leading dims while the second one unflattens them. Consider the following fake fx graph:

```
buf_0 = allgather(...)
buf_1 = aten.reshape(buf_0, ...)
buf_2 = aten.mm(buf_1, ...)
buf_3 = aten.reshape(buf_2, ...)
```

Since `fused_all_gather_matmul` semantically performs `matmul`s (as opposed to `mm`s), its results will replace `buf_0` and `buf_3`. It's okay if `buf_1` ends up with non-zero users, since it's just a view on `buf_0`. However, if for some reason `buf_2` ends up with non-zero users and can't be removed after the fusion (e.g. `buf_2` being returned for some reason), we'd be performing an extra `mm`.

To ensure `buf_2` has zero users after the fusion, since `buf_3` is always available and it's a reshape from `buf_2`, we replace `buf_2` with the reverse reshape from `buf_3`.